### ISSUE TITLE
Incorrect error about lifetimes with Send bound on AsyncFnMut::CallRefFuture

### ISSUE THREAD

**Issue by jacobsa** [rust-lang/rust#153558]
This program builds fine ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a47b40da3a1bd7d5f97a6d5d36359d21)):

```rust
/// Provides zero to an async function.
fn provide_zero<'a, F>(f: &'a mut F) -> impl Future<Output = ()> + 'a
where
    F: AsyncFnMut(i32),
{
    async { f(0).await }
}

/// Builds a vector containing zero in a really silly way.
async fn foo() -> Vec<i32> {
    let mut v = Vec::new();
    provide_zero(&mut async |n| v.push(n)).await;
    v
}
```

However, if you change `provide_zero` to require that the closure it accepts returns a future that can be polled in a multi-threaded environment, it stops compiling, with an incorrect error about `v` not living long enough in `foo` ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a040b4b91f290e9e18fe233b3920dda6)):

```rust
fn provide_zero<'a, F>(f: &'a mut F) -> impl Future<Output = ()> + 'a
where
    F: AsyncFnMut(i32),
    for<'b> F::CallRefFuture<'b>: Send,
{
    async { f(0).await }
}
```

```
   Compiling playground v0.0.1 (/playground)
error[E0597]: `v` does not live long enough
  --> src/lib.rs:18:23
   |
18 |     provide_zero(&mut async |n| v.push(n)).await;
   |     ------------------^^^^^^^^^^^^^^^^^^^-
   |     |                 |
   |     |                 borrowed value does not live long enough
   |     argument requires that `v` is borrowed for `'static`
19 |     v
20 | }
   | - `v` dropped here while still borrowed
   |
note: due to a current limitation of the type system, this implies a `'static` lifetime
  --> src/lib.rs:10:5
   |
10 |     for<'b> F::CallRefFuture<'b>: Send,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

From the note it seems clear this is a consequence of some known issue. But what known issue? What are the prospects for resolving it? (Googling the note only turns up #150081, which doesn't really tell me the answer.)

**Comment by eggyal** [comment]
The current limitation is that the generic associated type [`AsyncFnMut::CallRefFuture<'a>`](https://doc.rust-lang.org/stable/std/ops/trait.AsyncFnMut.html#associatedtype.CallRefFuture) requires `Self: 'a` and, since in your `provide_zero` function that generic lifetime takes the higher ranked trait bound `for<'b>`, you are requiring that any borrows in `F` outlive every possible lifetime `'b` — including `'static`, which the borrow of `v` clearly does not and cannot.

In this case, you simply need to use `'a` instead of the HRTB ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7322efc3b88caeafd5b59fed358cd9bc)):

```rust
fn provide_zero<'a, F>(f: &'a mut F) -> impl Future<Output = ()> + 'a
where
    F: AsyncFnMut(i32),
    F::CallRefFuture<'a>: Send,
{
    async { f(0).await }
}
```

Perhaps the diagnostic could be improved a little here, but otherwise this sort of question is probably better suited to the [Users' Forum](https://users.rust-lang.org).

@rustbot label +A-higher-ranked +A-GATs +D-newcomer-roadblock +D-papercut +T-types -needs-triage

**Comment by jacobsa** [comment]
Also, I'm not able to figure out a workaround for a more realistic example. I posted [this thread](https://users.rust-lang.org/t/how-to-teach-the-compiler-that-asyncfnmut-callreffuture-is-send-for-all-lifetimes-up-to-an-upper-bound/138786) on the users forum if anybody is able to help out.
